### PR TITLE
Cache bootstrap step in all CI workflows

### PR DESCRIPTION
**Pull request by arkq**
### Problem

Bootstrap takes ~5-6min to complete and is used in almost all workflows. Caching this step can save ~4-5min (restore takes ~1min).

### Changes

- moved all bootstrap configuration files to a dedicated directory (the hash of these files will be used as a cache key)
- added cache step before bootstrap step in workflow files
- use action cache feature for caching PyPi files (this will generate proper cache key - based on requirements content)

### Testing

CI will test that in a 2 steps process:
1. PR check should pass during initial check
2. PR check should pass after merging master changes (on second run, cache should be used for bootstrap step)

### CODE REVIEW

File: .github/workflows/build.yaml [hunk 1]
```diff
@@ -66,6 +66,14 @@ jobs:
               run: |
                   mkdir /tmp/cores || true
                   sysctl -w kernel.core_pattern=/tmp/cores/core.%u.%p.%t || true
+
+            - name: Bootstrap cache
+              uses: actions/cache@v3
+              with:
+                  key: ${{ runner.os }}-env-${{ hashFiles('scripts/setup/*') }}
```
Comment (andy31415) [inline .github/workflows/build.yaml:73]: bootstrap also depends on pigweed version. Is there some way to include that in the cache logic?

File: scripts/setup/bootstrap.sh [hunk 2]
```diff
@@ -0,0 +1,110 @@
+#
```
Comment (bzbarsky-apple) [inline scripts/setup/bootstrap.sh:1]: Did this just get moved with changes to  `s|scripts/|scripts/setup/|` and git's move detection failed somehow?  If not, what are the other changes here?

Comment (arkq) [reply scripts/setup/bootstrap.sh:1]: Yes it was moved to `/scripts/setup/`, but in order to maintain backward compatibility there is a symlink in `/scripts/`, which confuses git about rename... I would like to make it a rename (so the history will be preserver), but I don't know how to do this in a single PR... Normally, I would rename in one commit, and add symlink in the other.

File: .github/workflows/build.yaml (hunk 1 above)
Comment (arkq) [reply .github/workflows/build.yaml]: I've added pigweed repo dir files as a source for the hash: `hashFiles('scripts/setup/*', 'third_party/pigweed/**')`

**Review by andy31415 — APPROVED** [review]

**Review by bzbarsky-apple — APPROVED** [review]

**Pull request merged**